src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,287 @@
+
+  def seqOp[T <: Vector[Double]](sa: SkatAgg[T], st: SkatTuple[T]): SkatAgg[T] =
+    SkatAgg[T](sa.q + st.q, sa.xws + st.xw, sa.qtxws + st.qtxw)
+
+  def combOp[T <: Vector[Double]](sa: SkatAgg[T], sa2: SkatAgg[T]): SkatAgg[T] =
+    SkatAgg[T](sa.q + sa2.q, sa.xws ++ sa2.xws.result(), sa.qtxws ++ sa2.qtxws.result())
+
+  def denseResultOp(sa: SkatAgg[DenseVector[Double]], sigmaSq: Double): SkatStat = {
+
+    val m = sa.xws.length
+    val n = sa.xws(0).size
+    val k = sa.qtxws(0).length
+
+    var xwArray = new Array[Double](m * n)
+    var qtxwArray = new Array[Double](m * k)
+    var j = 0; var i = 0
+
+    //copy in non-zeros to weighted genotype matrix array
+    i = 0
+    while (i < m) {
+      j = 0
+      val xwsi = sa.xws(i)
+      while (j < n) {
+        xwArray(i * n + j) = xwsi(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //add in non-zeros to QtGW array
+    i = 0
+    while (i < m) {
+      j = 0
+      val qtxwsi = sa.qtxws(i)
+      while (j < k) {
+        qtxwArray(i * k + j) = qtxwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    val weightedGenotypes   = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t*weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t*qtWeightedGenotypes
+
+    val SPG = new SkatModel(GwGrammian, QtGwGrammian, sa.q / (2 *  sigmaSq))
+    SPG.computeSkatStats()
+  }
+
+  def sparseResultOp(sa: SkatAgg[SparseVector[Double]], sigmaSq: Double): SkatStat = {
+
+    val m = sa.xws.length
+    val n = sa.xws(0).size
+    val k = sa.qtxws(0).length
+
+    val xwArray   = new Array[Double](m * n)
+    val qtxwArray = new Array[Double](m * k)
+
+    var j = 0; var i = 0
+
+    while (i < m) {
+      val nnz = sa.xws(i).used
+      val xwsi = sa.xws(i)
+      j = 0
+      while (j < nnz) {
+        val index = sa.xws(i).index(j)
+        xwArray (i * n + index) = xwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //add in non-zeros to QtGW array
+    i = 0
+    while (i < m) {
+      j = 0
+      val qtxwsi = sa.qtxws(i)
+      while (j < k) {
+        qtxwArray(i * k + j) = qtxwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    val weightedGenotypes   = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t*weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t*qtWeightedGenotypes
+
+    val SPG = new SkatModel(GwGrammian, QtGwGrammian, sa.q / (2 *  sigmaSq))
+    SPG.computeSkatStats()
+  }
+
+  def largeNResultOp[T <: Vector[Double]](sa: SkatAgg[T], sigmaSq: Double): SkatStat = {
+    val m = sa.xws.length
+    val n = sa.xws(0).size
+    val k = sa.qtxws(0).length
+
+    val ZGrammianArray   = new Array[Double](m * m)
+    val QtZGrammianArray = new Array[Double](m * m)
+
+    var i = 0
+    var j = 0
+
+    while(i < m) {
```

Comment:
How should this be properly formatted? Should the variables not have spaces between their declaration and the while loop?